File: wbsviz/core/csv_reader.py
"""
CSV reader for WBS data files.
"""

from typing import List
from pathlib import Path
from wbsviz.core.model import Record
from wbsviz.core.common import split_csv_line, to_lower


class CsvReader:
    """Reads and parses CSV files containing WBS data."""
    
    def __init__(self, path: str):
        """
        Initialize CSV reader.
        
        Args:
            path: Path to CSV file
        """
        self.path = Path(path)
    
    def read(self) -> List[Record]:
        """
        Read and parse CSV file.
        
        Returns:
            List of Record objects
            
        Raises:
            FileNotFoundError: If file doesn't exist
            ValueError: If CSV is invalid or missing required fields
        """
        if not self.path.exists():
            raise FileNotFoundError(f"Failed to open CSV: {self.path}")
        
        with open(self.path, 'r', encoding='utf-8') as f:
            # Read header
            header_line = f.readline()
            if not header_line.strip():
                raise ValueError("Empty CSV")
            
            headers = split_csv_line(header_line.strip())
            headers_lower = [to_lower(h) for h in headers]
            
            # Find column indices
            def find_index(key: str) -> int:
                key_lower = to_lower(key)
                for i, h in enumerate(headers_lower):
                    if h == key_lower:
                        return i
                return -1
            
            # Required headers
            code_idx = find_index("code")
            title_idx = find_index("title")
            
            if code_idx < 0 or title_idx < 0:
                raise ValueError("CSV missing required headers: code, title")
            
            # Optional headers
            desc_idx = find_index("description")
            pr_idx = find_index("primaryresp")
            sr_idx = find_index("seconderyresp")
            est_idx = find_index("estimateduration")
            
            # Read data rows
            records = []
            line_num = 1  # Header is line 1
            
            for line in f:
                line_num += 1
                line = line.strip()
                if not line:
                    continue
                
                row = split_csv_line(line)
                
                def get_field(idx: int) -> str:
                    if 0 <= idx < len(row):
                        return row[idx]
                    return ""
                
                code = get_field(code_idx)
                title = get_field(title_idx)
                
                if not code or not title:
                    raise ValueError(f"Missing code/title at line {line_num}")
                
                record = Record(
                    code=code,
                    title=title,
                    description=get_field(desc_idx),
                    primaryResp=get_field(pr_idx),
                    seconderyResp=get_field(sr_idx),
                    estimateDuration=get_field(est_idx)
                )
                
                records.append(record)
            
            return records

File: legacy/src/CsvReader.cpp
#include "wbsviz/CsvReader.hpp"
#include "wbsviz/Common.hpp"
#include <fstream>
#include <stdexcept>

namespace wbsviz {

CsvReader::CsvReader(std::string path) : path_(std::move(path)) {}

std::vector<Record> CsvReader::read() const {
    std::ifstream in(path_);
    if (!in) throw std::runtime_error("Failed to open CSV: " + path_);

    std::string hdr; if(!std::getline(in,hdr)) throw std::runtime_error("Empty CSV");
    auto H = splitCsvLine(hdr);
    for (auto& h: H) h = toLower(h);  // case-insensitive headers

    auto idx = [&](const std::string& k)->int{
        auto kl = toLower(k);
        for (size_t i=0;i<H.size();++i) if (H[i]==kl) return static_cast<int>(i);
        return -1;
    };

    // Required headers
    const int i_code  = idx("code");
    const int i_title = idx("title");
    if (i_code<0 || i_title<0)
        throw std::runtime_error("CSV missing required headers: code,title");

    //Optional
    const int i_desc  = idx("description");
    const int i_pr    = idx("primaryresp");
    const int i_sr    = idx("seconderyresp");   
    const int i_est   = idx("estimateduration");

    std::vector<Record> out; std::string line; size_t ln=1;
    while (std::getline(in,line)) {
        ++ln; if(line.empty()) continue;
        auto row = splitCsvLine(line);
        auto get = [&](int i)->std::string { return (i>=0 && i < (int)row.size()) ? row[i] : std::string{}; };

        Record r;
        r.code            = get(i_code);
        r.title           = get(i_title);
        r.description     = get(i_desc);
        r.primaryResp     = get(i_pr);
        r.seconderyResp   = get(i_sr);
        r.estimateDuration= get(i_est);

        if (r.code.empty() || r.title.empty())
            throw std::runtime_error("Missing code/title at line " + std::to_string(ln));

        out.push_back(std::move(r));
    }
    return out;
}

} // namespace wbsviz
